File: firefox/src/java/com/googlecode/webdriver/firefox/internal/ExtensionConnectionFactory.java
package com.googlecode.webdriver.firefox.internal;

import com.googlecode.webdriver.firefox.ExtensionConnection;
import com.googlecode.webdriver.firefox.FirefoxProfile;

import java.io.File;
import java.io.IOException;

public class ExtensionConnectionFactory {
    public static ExtensionConnection connectTo(FirefoxProfile profile, String host, int port) {
        boolean isDev = Boolean.getBoolean("webdriver.firefox.useExisting");
        if (isDev) {
            try {
                return new RunningInstanceConnection(host, port);
            } catch (IOException e) {
                // Fine. No running instance
            }
        }

        try {
          return new NewProfileExtensionConnection(profile, host, port);
        } catch (Exception e) {
            // Then we can't connect
        }

        return new DisconnectedExtension();
    }
}

File: firefox/src/csharp/webdriver-firefox/Internal/ExtensionConnectionFactory.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace OpenQA.Selenium.Firefox.Internal
{
    /// <summary>
    /// Creates connections to a running instance of the WebDriver Firefox extension.
    /// </summary>
    internal static class ExtensionConnectionFactory
    {
        /// <summary>
        /// Connect to an instance of the WebDriver Firefox extension.
        /// </summary>
        /// <param name="binary">The <see cref="FirefoxBinary"/> in which the extension is hosted.</param>
        /// <param name="profile">The <see cref="FirefoxProfile"/> in which the extension is installed.</param>
        /// <param name="host">The host name of the computer running the extension (usually "localhost").</param>
        /// <returns>An <see cref="ExtensionConnection"/> connected to the WebDriver Firefox extension.</returns>
        /// <exception cref="WebDriverException">If there are any problems connecting to the extension.</exception>
        public static ExtensionConnection ConnectTo(FirefoxBinary binary, FirefoxProfile profile, string host)
        {
            int profilePort = profile.Port;
            ExtensionConnection connection = null;
            using (ILock lockObject = new SocketLock(profilePort - 1))
            {
                connection = new ExtensionConnection(lockObject, binary, profile, host);
            }

            return connection;
        }
    }
}
